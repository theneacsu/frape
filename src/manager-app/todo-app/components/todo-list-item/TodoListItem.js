import React, { Component } from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { connect } from 'react-redux'
import TodoForm from '../todo-form/TodoForm'
import {
  startRemoveTodo,
  startEditTodo,
  startToggleTodo
} from '../../../../actions/todo-app/todos'
import { getTodoById } from '../../../../selectors/todo-app/todos'
import styles from './TodoListItem.module.css'

export class TodoListItem extends Component {
  state = {
    editMode: false,
    isCompleted: this.props.completed
  }

  onSubmitHandler = todo => {
    this.props.startEditTodo(this.props.id, todo)
    this.setState(() => ({ editMode: false }))
  }

  onToggleHandler = id => {
    this.props.startToggleTodo(id)
    this.setState(prevState => ({
      isCompleted: !prevState.isCompleted
    }))
  }

  render() {
    const todoListItemClasses = this.state.isCompleted
      ? [styles.textDiv, styles.completed].join(' ')
      : styles.textDiv
    let todoListItem = this.state.editMode ? (
      <TodoForm
        buttonText="Save"
        todo={this.props.todo}
        onSubmit={this.onSubmitHandler}
      />
    ) : (
      <div className={styles.wrapperDiv}>
        <div className={todoListItemClasses}>
          <label className={styles.label}>
            <div className={styles.round}>
              <input
                className={styles.checkbox}
                type="checkbox"
                checked={this.props.completed}
                onChange={() => this.onToggleHandler(this.props.id)}
              />
              <label onClick={() => this.onToggleHandler(this.props.id)} />
            </div>

            <span className={styles.span}>{this.props.text}</span>
          </label>
        </div>
        <div className={styles.iconsDiv}>
          <FontAwesomeIcon
            className={styles.icon}
            icon="pencil-alt"
            color="#6347FF"
            onClick={() =>
              this.setState(prevState => ({ editMode: !prevState.editMode }))
            }
          />
          <FontAwesomeIcon
            className={styles.icon}
            icon="trash-alt"
            color="#FF6347"
            onClick={() => this.props.startRemoveTodo(this.props.id)}
          />
        </div>
      </div>
    )

    return todoListItem
  }
}

const mapDispatchToProps = (dispatch, props) => ({
  startRemoveTodo: id => dispatch(startRemoveTodo(id)),
  startEditTodo: (id, updates) => dispatch(startEditTodo(id, updates)),
  startToggleTodo: id => dispatch(startToggleTodo(id))
})

const mapStateToProps = (state, props) => ({
  todo: getTodoById(state.todos, props.id)
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TodoListItem)
