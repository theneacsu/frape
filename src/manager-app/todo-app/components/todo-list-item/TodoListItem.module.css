.wrapperDiv {
  display: flex;
  justify-content: space-between;
  margin: 1rem 0;
  border: 1px solid #4cb69f;
  border-radius: 2rem;
  padding: 2rem;
}

.textDiv {
  display: flex;
  justify-content: flex-start;
  align-items: center;
}

.label {
  display: flex;
  align-items: center;
}

.span {
  text-align: left;
  margin-left: 3rem;
  display: inline-block;
  max-width: 50vw;
  word-wrap: break-word;
}
.span:hover {
  cursor: pointer;
}

.iconsDiv {
  width: 70px;
  display: flex;
  vertical-align: middle;
  align-items: center;
}

.icon {
  margin-left: 2rem;
}

.icon:hover {
  cursor: pointer;
}

.completed {
  text-decoration: line-through;
}

.round {
  position: relative;
}

.round label {
  background-color: #fff;
  border: 1px solid #ccc;
  border-radius: 50%;
  cursor: pointer;
  height: 26px;
  left: 0;
  position: absolute;
  top: 0;
  width: 26px;
}

.round label:after {
  border: 2px solid #fff;
  border-top: none;
  border-right: none;
  content: '';
  height: 4px;
  left: 5px;
  opacity: 0;
  position: absolute;
  top: 8px;
  transform: rotate(-45deg);
  width: 12px;
}

.round input[type='checkbox'] {
  visibility: hidden;
}

.round input[type='checkbox']:checked + label {
  background-color: #4cb69f;
  border-color: #4cb69f;
}

.round input[type='checkbox']:checked + label:after {
  opacity: 1;
}
