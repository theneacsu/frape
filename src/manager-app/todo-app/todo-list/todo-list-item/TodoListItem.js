import React, { Component } from 'react'
import { connect } from 'react-redux'
import TodoForm from '../../todo-form/TodoForm'
import {
  removeTodo,
  editTodo,
  toggleTodo
} from '../../../../actions/todo-app/todos'
import { getTodoById } from '../../../../selectors/todo-app/todos'

export class TodoListItem extends Component {
  state = {
    editMode: false
  }

  onSubmitHandler = todo => {
    this.props.editTodo(this.props.id, todo)
    this.setState(() => ({ editMode: false }))
  }

  render() {
    let todoListItem = this.state.editMode ? (
      <TodoForm
        buttonText="Save"
        todo={this.props.todo}
        onSubmit={this.onSubmitHandler}
      />
    ) : (
      <div>
        <label>
          <input
            type="checkbox"
            onChange={() => this.props.toggleTodo(this.props.id)}
          />
          <span>{this.props.text}</span>
        </label>
        <button
          onClick={() =>
            this.setState(prevState => ({ editMode: !prevState.editMode }))
          }
        >
          Edit
        </button>
        <button onClick={() => this.props.removeTodo(this.props.id)}>
          Remove
        </button>
      </div>
    )

    return todoListItem
  }
}

const mapDispatchToProps = (dispatch, props) => ({
  removeTodo: id => dispatch(removeTodo(id)),
  editTodo: (id, updates) => dispatch(editTodo(id, updates)),
  toggleTodo: id => dispatch(toggleTodo(id))
})

const mapStateToProps = (state, props) => ({
  todo: getTodoById(state.todos, props.id)
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TodoListItem)
