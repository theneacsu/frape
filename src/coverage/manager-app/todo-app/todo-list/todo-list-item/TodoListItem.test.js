import React from 'react'
import { shallow } from 'enzyme'
import { TodoListItem } from '../../../../../manager-app/todo-app/todo-list/todo-list-item/TodoListItem'

let todo, removeTodoSpy, editTodoSpy, toggleTodoSpy, wrapper

beforeEach(() => {
  todo = {
    text: 'go for a run',
    completed: true
  }
  removeTodoSpy = jest.fn()
  editTodoSpy = jest.fn()
  toggleTodoSpy = jest.fn()
  wrapper = shallow(
    <TodoListItem
      removeTodo={removeTodoSpy}
      editTodo={editTodoSpy}
      toggleTodo={toggleTodoSpy}
      todo={todo}
    />
  )
})

test('Should render correctly the TodoListItem component', () => {
  expect(wrapper).toMatchSnapshot()
})

test('Should remove the todo', () => {
  wrapper
    .find('button')
    .at(1)
    .simulate('click')
  expect(removeTodoSpy).toHaveBeenLastCalledWith(wrapper.prop('id'))
})

test('Should toggle the todo completed status', () => {
  wrapper
    .find('input')
    .at(0)
    .simulate('change')
  expect(toggleTodoSpy).toHaveBeenLastCalledWith(wrapper.prop('id'))
})

test('Should edit the todo', () => {
  wrapper
    .find('button')
    .at(0)
    .simulate('click')
  expect(wrapper.state('editMode')).toBe(true)
  wrapper.find('TodoForm').prop('onSubmit')(todo)
  expect(editTodoSpy).toHaveBeenLastCalledWith(wrapper.prop('id'), todo)
})
