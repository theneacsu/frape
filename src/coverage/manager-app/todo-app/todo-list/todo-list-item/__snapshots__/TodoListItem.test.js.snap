// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correctly the TodoListItem component 1`] = `
<div>
  <label>
    <input
      onChange={[Function]}
      type="checkbox"
    />
    <span />
  </label>
  <button
    onClick={[Function]}
  >
    Edit
  </button>
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
